src/rgw/rgw_data_sync.cc

```diff
@@ -194,6 +232,50 @@ int RGWReadDataSyncStatusCoroutine::operate()
   return 0;
 }
 
+class RGWReadRecoveringShardsCoroutine : public RGWCoroutine {
+  RGWDataSyncEnv *sync_env;
+  rgw_data_sync_status *sync_status;
+  set<int>& recovering_shards;
+
+  map<int, std::set<std::string>> entries_map;
+  int max_entries{1};
+
+public:
+  RGWReadRecoveringShardsCoroutine(RGWDataSyncEnv *_sync_env,
+                                   rgw_data_sync_status *_status, set<int>& _recovering_shards)
+    : RGWCoroutine(_sync_env->cct), sync_env(_sync_env), sync_status(_status), recovering_shards(_recovering_shards)
+  {}
+  int operate() override;
+};
+
+int RGWReadRecoveringShardsCoroutine::operate()
```

Comment:
since this coroutine is only making one asynchronous call to `RGWReadDataSyncRecoveringShardsCR`, with some extra logic for setup/teardown, it probably makes sense to move that extra logic into `read_recovering_shards()` so we only need the one `RGWReadDataSyncRecoveringShardsCR` coroutine